### PR TITLE
CSHARP-4412: BsonRepresentationAttribute not respected when string array assigned to IList property.

### PR DESCRIPTION
**Pull request by rstam**

### CODE REVIEW

File: src/MongoDB.Bson/Serialization/CollectionsSerializationProvider.cs [hunk 1]
```diff
@@ -232,12 +232,11 @@ private IBsonSerializer GetCollectionSerializer(Type type, IBsonSerializerRegist
...
                 else if (typeInfo.IsInterface)
                 {
-                    var listDefinition = typeof(List<>);
-                    var listType = listDefinition.MakeGenericType(itemType);
+                    var listType = typeof(List<>).MakeGenericType(itemType);
                     if (typeInfo.IsAssignableFrom(listType))
                     {
-                        var serializerDefinition = typeof(ImpliedImplementationInterfaceSerializer<,>);
-                        return CreateGenericSerializer(serializerDefinition, new[] { type, listType }, serializerRegistry);
+                        var serializerDefinition = typeof(IEnumerableListSerializer<,>);
```
Comment (rstam) [inline src/MongoDB.Bson/Serialization/CollectionsSerializationProvider.cs:238]: Use the new serializer for `IEnumerable` (and its subinterfaces that List implements).

Comment (rstam) [reply src/MongoDB.Bson/Serialization/CollectionsSerializationProvider.cs:238]: The problem with `ImpliedImplementationInterfaceSerializer` is that the `[BsonRepresentation]` attribute only worked when the actual type was the same as the implied implementation type.

**Review by JamesKovacs — APPROVED** [review]
LGTM

File: tests/MongoDB.Bson.Tests/Jira/CSharp4412Tests.cs [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+        }
+
+        [Fact]
+        public void ISet_Ids_should_deserialize_from_ObjectIds()
+        {
+            var json = "{ \"Ids\" : [ObjectId(\"0102030405060708090a0b0c\")] }";
+
+            var rehydrated = BsonSerializer.Deserialize<ClassWithISetIds>(json);
+
+            rehydrated.Ids.Should().Equal("0102030405060708090a0b0c");
```
Comment (BorisDog) [inline tests/MongoDB.Bson.Tests/Jira/CSharp4412Tests.cs:125]: It there a value in checking `typeof(rehydrated.Ids) == Hashset<>` ? Or do we want to have flexibility of concrete implementation?

File: src/MongoDB.Bson/Serialization/Serializers/IEnumerableDeserializingAsListSerializer.cs [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+        IChildSerializerConfigurable
+        where TIEnumerable : class, IEnumerable<TItem> // TIEnumerable must be an interface
+    {
+        #region static
+        private static void EnsureTIEnumerableIsAnInterface()
+        {
+            if (!typeof(TIEnumerable).IsInterface)
+            {
+                // this constraint cannot be specified at compile time
+                throw new ArgumentException($"The TIEnumerable type argument is not an interface: {typeof(TIEnumerable)}.", nameof(TIEnumerable));
```
[the remark this answers is not in the document]
Comment (rstam) [reply src/MongoDB.Bson/Serialization/Serializers/IEnumerableDeserializingAsListSerializer.cs]: Done.

File: src/MongoDB.Bson/Serialization/Serializers/IEnumerableDeserializingAsListSerializer.cs [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+    public class IEnumerableDeserializingAsListSerializer<TIEnumerable, TItem> :
+        SerializerBase<TIEnumerable>,
+        IBsonArraySerializer,
+        IChildSerializerConfigurable
+        where TIEnumerable : class, IEnumerable<TItem> // TIEnumerable must be an interface
+    {
+        #region static
+        private static void EnsureTIEnumerableIsAnInterface()
+        {
+            if (!typeof(TIEnumerable).IsInterface)
```
[the remark this answers is not in the document]
Comment (rstam) [reply src/MongoDB.Bson/Serialization/Serializers/IEnumerableDeserializingAsListSerializer.cs]: Discussed in Slack. Throwing exceptions from static constructors is not recommended. They can get thrown at unpredictable times which can make it harder for the application developer to figure out where in their own code the problem is.

File: src/MongoDB.Bson/Serialization/Serializers/IEnumerableDeserializingAsListSerializer.cs [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the IEnumerableDeserializingAsListSerializer class.
+        /// </summary>
+        /// <param name="serializerRegistry">The serializer registry.</param>
+        public IEnumerableDeserializingAsListSerializer(IBsonSerializerRegistry serializerRegistry)
+        {
+            EnsureTIEnumerableIsAnInterface();
+            if (serializerRegistry == null)
```
[the remark this answers is not in the document]
Comment (rstam) [reply src/MongoDB.Bson/Serialization/Serializers/IEnumerableDeserializingAsListSerializer.cs]: `Ensure.IsNotNull` is not available in the `Bson` layer.

It probably should be, but currently is not.

File: src/MongoDB.Bson/Serialization/Serializers/IEnumerableDeserializingAsListSerializer.cs [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+        }
+
+        // public properties
+        /// <summary>
+        /// Gets the item serializer.
+        /// </summary>
+        /// <value>
+        /// The item serializer.
+        /// </value>
+        public IBsonSerializer<TItem> ItemSerializer
```
[the remark this answers is not in the document]
Comment (rstam) [reply src/MongoDB.Bson/Serialization/Serializers/IEnumerableDeserializingAsListSerializer.cs]: Agreed. I used an existing class as the basis for this class, and the original code was written a long time ago.

[the remark this answers is not in the document]
Comment (rstam) [reply src/MongoDB.Bson/Serialization/Serializers/IEnumerableDeserializingAsListSerializer.cs]: Done.

File: tests/MongoDB.Bson.Tests/Jira/CSharp4412Tests.cs (hunk 2 above)
Comment (rstam) [reply tests/MongoDB.Bson.Tests/Jira/CSharp4412Tests.cs:125]: I wondered about that also.

But decided it didn't matter what the actual type was. It only mattered that it implemented the interface.

File: src/MongoDB.Bson/Serialization/Serializers/IEnumerableDeserializingAsCollectionSerializer.cs [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+
+namespace MongoDB.Bson.Serialization.Serializers
+{
+    /// <summary>
+    /// Represents a serializer for IEnumerable and any other derived interface implemented by TCollection.
+    /// </summary>
+    /// <typeparam name="TIEnumerable">The type of an IEnumerable interface.</typeparam>
+    /// <typeparam name="TItem">The type of the items.</typeparam>
+    /// <typeparam name="TCollection">The type of the collection used when deserializing.</typeparam>
+    public class IEnumerableDeserializingAsCollectionSerializer<TIEnumerable, TItem, TCollection> :
```
Comment (rstam) [inline src/MongoDB.Bson/Serialization/Serializers/IEnumerableDeserializingAsCollectionSerializer.cs:27]: Combining the two new serializers into a single one worked out well. Thanks for the suggestion.

Note that the new type parameter is `TCollection`, not `TICollection`

File: src/MongoDB.Bson/Serialization/Serializers/IEnumerableDeserializingAsCollectionSerializer.cs [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+        /// </summary>
+        /// <param name="itemSerializer">The item serializer.</param>
+        /// <returns>The reconfigured serializer.</returns>
+        public IEnumerableDeserializingAsCollectionSerializer<TIEnumerable, TItem, TCollection> WithItemSerializer(IBsonSerializer<TItem> itemSerializer)
+        {
+            return new IEnumerableDeserializingAsCollectionSerializer<TIEnumerable, TItem, TCollection>(itemSerializer);
+        }
+
+        // explicit interface implementations
+        IBsonSerializer IChildSerializerConfigurable.ChildSerializer
```
Comment (BorisDog) [inline src/MongoDB.Bson/Serialization/Serializers/IEnumerableDeserializingAsCollectionSerializer.cs:168]: Expression body?

File: src/MongoDB.Bson/Serialization/Serializers/IEnumerableDeserializingAsCollectionSerializer.cs [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+        {
+            get { return ItemSerializer; }
+        }
+
+        IBsonSerializer IChildSerializerConfigurable.WithChildSerializer(IBsonSerializer childSerializer)
```
Comment (BorisDog) [inline src/MongoDB.Bson/Serialization/Serializers/IEnumerableDeserializingAsCollectionSerializer.cs:173]: Expression body?

File: src/MongoDB.Bson/Serialization/Serializers/IEnumerableDeserializingAsCollectionSerializer.cs [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+        /// <summary>
+        /// Initializes a new instance of the IEnumerableDeserializingAsCollectionSerializer class.
+        /// </summary>
+        /// <param name="serializerRegistry">The serializer registry.</param>
+        public IEnumerableDeserializingAsCollectionSerializer(IBsonSerializerRegistry serializerRegistry)
+        {
+            EnsureTIEnumerableIsAnInterface();
+            if (serializerRegistry == null)
+            {
+                throw new ArgumentNullException("serializerRegistry");
```
Comment (BorisDog) [inline src/MongoDB.Bson/Serialization/Serializers/IEnumerableDeserializingAsCollectionSerializer.cs:81]: nameof(serializerRegistry)

File: src/MongoDB.Bson/Serialization/Serializers/IEnumerableDeserializingAsCollectionSerializer.cs [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+        /// <summary>
+        /// Initializes a new instance of the IEnumerableDeserializingAsCollectionSerializer class.
+        /// </summary>
+        /// <param name="itemSerializer">The item serializer.</param>
+        public IEnumerableDeserializingAsCollectionSerializer(IBsonSerializer<TItem> itemSerializer)
+        {
+            EnsureTIEnumerableIsAnInterface();
+            if (itemSerializer == null)
+            {
+                throw new ArgumentNullException("itemSerializer");
```
Comment (BorisDog) [inline src/MongoDB.Bson/Serialization/Serializers/IEnumerableDeserializingAsCollectionSerializer.cs:66]: nameof(itemSerializer)

File: src/MongoDB.Bson/Serialization/Serializers/IEnumerableDeserializingAsCollectionSerializer.cs [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+            }
+
+            _lazyItemSerializer = new Lazy<IBsonSerializer<TItem>>(() => itemSerializer);
```
Comment (BorisDog) [inline src/MongoDB.Bson/Serialization/Serializers/IEnumerableDeserializingAsCollectionSerializer.cs:69]: I wish we could use  `Lazy<T>(T value)` overload available in .net standard 2.1

File: src/MongoDB.Bson/Serialization/Serializers/IEnumerableDeserializingAsCollectionSerializer.cs [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+        /// <param name="serializerRegistry">The serializer registry.</param>
+        public IEnumerableDeserializingAsCollectionSerializer(IBsonSerializerRegistry serializerRegistry)
+        {
+            EnsureTIEnumerableIsAnInterface();
+            if (serializerRegistry == null)
+            {
+                throw new ArgumentNullException("serializerRegistry");
+            }
+
+            _lazyItemSerializer = new Lazy<IBsonSerializer<TItem>>(() => serializerRegistry.GetSerializer<TItem>());
```
Comment (BorisDog) [inline src/MongoDB.Bson/Serialization/Serializers/IEnumerableDeserializingAsCollectionSerializer.cs:84]: looks like we can skip the additional lambda creation: 
`new Lazy<IBsonSerializer<TItem>>(serializerRegistry.GetSerializer<TItem>);`

**Review by BorisDog — COMMENTED** [review]
Few minor comments.

File: src/MongoDB.Bson/Serialization/Serializers/IEnumerableDeserializingAsCollectionSerializer.cs (hunk 11 above)
Comment (rstam) [reply src/MongoDB.Bson/Serialization/Serializers/IEnumerableDeserializingAsCollectionSerializer.cs:66]: Done

File: src/MongoDB.Bson/Serialization/Serializers/IEnumerableDeserializingAsCollectionSerializer.cs (hunk 12 above)
Comment (rstam) [reply src/MongoDB.Bson/Serialization/Serializers/IEnumerableDeserializingAsCollectionSerializer.cs:69]: True

File: src/MongoDB.Bson/Serialization/Serializers/IEnumerableDeserializingAsCollectionSerializer.cs (hunk 10 above)
Comment (rstam) [reply src/MongoDB.Bson/Serialization/Serializers/IEnumerableDeserializingAsCollectionSerializer.cs:81]: Done

File: src/MongoDB.Bson/Serialization/Serializers/IEnumerableDeserializingAsCollectionSerializer.cs (hunk 13 above)
Comment (rstam) [reply src/MongoDB.Bson/Serialization/Serializers/IEnumerableDeserializingAsCollectionSerializer.cs:84]: Done

File: src/MongoDB.Bson/Serialization/Serializers/IEnumerableDeserializingAsCollectionSerializer.cs (hunk 8 above)
Comment (rstam) [reply src/MongoDB.Bson/Serialization/Serializers/IEnumerableDeserializingAsCollectionSerializer.cs:168]: Done;

File: src/MongoDB.Bson/Serialization/Serializers/IEnumerableDeserializingAsCollectionSerializer.cs (hunk 9 above)
Comment (rstam) [reply src/MongoDB.Bson/Serialization/Serializers/IEnumerableDeserializingAsCollectionSerializer.cs:173]: I don't like expression bodies for methods.

But done anyway.

Comment (rstam) [reply src/MongoDB.Bson/Serialization/Serializers/IEnumerableDeserializingAsCollectionSerializer.cs:173]: FYI the reason I don't like expression bodies for methods is that most methods aren't one liners, and I find it visually disconcerting when some methods "appear" to be missing their body.

**Review by BorisDog — APPROVED** [review]
LGTM!

**Pull request merged**